src/libcore/str/pattern.rs

```diff
@@ -630,3 +509,420 @@ impl<'a, F> Pattern<'a> for F where F: FnMut(char) -> bool {
+        let byteset = needle.iter()
+                            .fold(0, |a, &b| (1 << ((b & 0x3f) as usize)) | a);
+
+        // A particularly readable explanation of what's going on here can be found
+        // in Crochemore and Rytter's book "Text Algorithms", ch 13. Specifically
+        // see the code for "Algorithm CP" on p. 323.
+        //
+        // What's going on is we have some critical factorization (u, v) of the
+        // needle, and we want to determine whether u is a suffix of
+        // &v[..period]. If it is, we use "Algorithm CP1". Otherwise we use
+        // "Algorithm CP2", which is optimized for when the period of the needle
+        // is large.
+        if &needle[..crit_pos] == &needle[period.. period + crit_pos] {
+            TwoWaySearcher {
+                crit_pos: crit_pos,
+                period: period,
+                byteset: byteset,
+
+                position: 0,
+                end: !0,
+                memory: 0
+            }
+        } else {
+            TwoWaySearcher {
+                crit_pos: crit_pos,
+                period: cmp::max(crit_pos, needle.len() - crit_pos) + 1,
+                byteset: byteset,
+
+                position: 0,
+                end: !0,
+                memory: usize::MAX // Dummy value to signify that the period is long
+            }
+        }
+    }
+
+    // One of the main ideas of Two-Way is that we factorize the needle into
+    // two halves, (u, v), and begin trying to find v in the haystack by scanning
+    // left to right. If v matches, we try to match u by scanning right to left.
+    // How far we can jump when we encounter a mismatch is all based on the fact
+    // that (u, v) is a critical factorization for the needle.
+    #[inline]
+    fn next(&mut self, haystack: &[u8], needle: &[u8], long_period: bool)
+            -> SearchStep {
+        // `next()` uses `self.position` as its cursor
+        let start = self.position;
+        'search: loop {
+            // Check that we have room to search in
+            if self.position + needle.len() > haystack.len() {
+                if start == haystack.len() {
+                    return SearchStep::Done
+                } else {
+                    self.position = haystack.len();
+                }
+            }
+            if start != self.position {
+                return SearchStep::Reject(start, self.position);
+            }
+
+            // Quickly skip by large portions unrelated to our substring
+            if (self.byteset >>
+                    ((haystack[self.position + needle.len() - 1] & 0x3f)
+                     as usize)) & 1 == 0 {
+                self.position += needle.len();
+                if !long_period {
+                    self.memory = 0;
+                }
+                continue 'search;
+            }
+
+            // See if the right part of the needle matches
+            let start = if long_period { self.crit_pos }
```

Comment:
Could the searching of the left and right parts be put into their own blocks? The shadowing of `start` seems very confusing, especially as the three instances of the same name refer to different things, and separate blocks would remove that shadowing.